Rename nvcf -> nvqc

1tnguyen opened the pull request:
<!--
Thanks for helping us improve CUDA Quantum!

⚠️ The pull request title should be concise and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

Checklist:
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Description
<!-- Include relevant issues here, describe what changed and why -->

Rename `nvcf` to `nvqc`.

**Note**: this PR is related to https://github.com/NVIDIA/cuda-quantum/pull/1227. After the first one merged in, we need to rebase the other.

bmhowe23 reviewed (commented):
It looks like there are still some references to NVCF in error messages in the `RemoteRestClient.cpp` (like "Please check your NVCF configurations", etc.). Since we are changing all the user-facing portions to NVQC (including items like `NVQC_API_KEY` and `NVQC_FUNCTION_ID`), we should probably update those error messages, too. (There's probably some judgement calls about whether some items in there should stay NVCF, but I would favor changing the ones that talk about NVCF configurations or API keys.)

schweitzpgi commented:
> It looks like there are still some references to NVCF in error messages in the `RemoteRestClient.cpp` (like "Please check your NVCF configurations", etc.). Since we are changing all the user-facing portions to NVQC (including items like `NVQC_API_KEY` and `NVQC_FUNCTION_ID`), we should probably update those error messages, too. (There's probably some judgement calls about whether some items in there should stay NVCF, but I would favor changing the ones that talk about NVCF configurations or API keys.)

Edification question: Does NVCF mean something distinctive? A marketing acronym?

bmhowe23 commented:
> > It looks like there are still some references to NVCF in error messages in the `RemoteRestClient.cpp` (like "Please check your NVCF configurations", etc.). Since we are changing all the user-facing portions to NVQC (including items like `NVQC_API_KEY` and `NVQC_FUNCTION_ID`), we should probably update those error messages, too. (There's probably some judgement calls about whether some items in there should stay NVCF, but I would favor changing the ones that talk about NVCF configurations or API keys.)
> 
> Edification question: Does NVCF mean something distinctive? A marketing acronym?

Yes, it refers to this: https://docs.nvidia.com/cloud-functions/index.html

1tnguyen commented:
> It looks like there are still some references to NVCF in error messages in the `RemoteRestClient.cpp` (like "Please check your NVCF configurations", etc.). Since we are changing all the user-facing portions to NVQC (including items like `NVQC_API_KEY` and `NVQC_FUNCTION_ID`), we should probably update those error messages, too. (There's probably some judgement calls about whether some items in there should stay NVCF, but I would favor changing the ones that talk about NVCF configurations or API keys.)

I've updated those error and log messages in `RemoteRestClient.cpp` to use NVQC as much as possible:  https://github.com/NVIDIA/cuda-quantum/pull/1247/commits/0c873be86e296da4b7f459462426adaa8ab764d4

The validation of API key and function Id would actually be hit earlier at the QPU level. Hence, I've updated the error message there to be more informative: https://github.com/NVIDIA/cuda-quantum/pull/1247/commits/e2aca78fe44ed7aaa611115be129bb773586c0ba 
Since there are a few input mechanisms, I only refer to the future docs in the error message.

bettinaheim commented on docs/sphinx/examples/cpp/providers/nvqc_sample.cpp:
````diff
@@ -1,23 +1,23 @@
 // ```
-// Assumes a valid NVCF API key and function ID have been stored in environment
-// variables or `~/.nvcf_config` file. Alternatively, they can be set in the
+// Assumes a valid NVQC API key and function ID have been stored in environment
````
Why does the function ID need to be stored as environment variable?
The CUDA Quantum runtime should take care of that automatically. Since this is a public example, I would remove any mention of the function-id configuration option - at this point, we do not expect that anyone would use a custom function, do we?

1tnguyen replied:
Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1247/commits/ae21a9b4165762420178cb7cbfa3a4cdf89892af:

- Query NVCF to get available functions. 

- Client to track the number of GPUs of each available deployed function. Add `ngpus` option.

- Update examples to get rid of function Id.

Side note: I figured that it's more convenient to implement these in this PR rather than a separate PR.

bettinaheim commented on docs/sphinx/examples/cpp/providers/nvqc_state.cpp:
```diff
@@ -1,17 +1,17 @@
-// nvq++ --target nvcf --nvcf-backend tensornet --nvcf-api-key <YOUR API KEY> \
-// --nvcf-function-id <NVCF function Id> nvcf_state.cpp -o out.x
+// nvq++ --target nvqc --nvqc-backend tensornet --nvqc-api-key <YOUR API KEY> \
+// --nvqc-function-id <NVQC function Id> nvqc_state.cpp -o out.x
```
Same comment here, and same for the python examples.

1tnguyen replied:
Fixed in https://github.com/NVIDIA/cuda-quantum/commit/ae21a9b4165762420178cb7cbfa3a4cdf89892af

bettinaheim commented on scripts/validate_container.sh:
```diff
@@ -131,7 +131,7 @@ do
     let "samples+=1"
 
-    if [[ "$ex" == *"iqm"* ]] || [[ "$ex" == *"ionq"* ]] || [[ "$ex" == *"quantinuum"* ]] || [[ "$ex" == *"nvcf"* ]];
+    if [[ "$ex" == *"iqm"* ]] || [[ "$ex" == *"ionq"* ]] || [[ "$ex" == *"quantinuum"* ]] || [[ "$ex" == *"nvqc"* ]];
```
We should think about reenabling pre-merge checks for the nvqc backend - let's chat about this offline.

1tnguyen replied:
Discussed offline. We'll enable pre-merge checks for NVQC once we have a semi-permanent/stable deployment.

bettinaheim commented on runtime/cudaq/platform/mqpu/nvqc.config:
```diff
@@ -27,19 +27,19 @@ PLATFORM_EXTRA_ARGS=""
 		PLATFORM_EXTRA_ARGS="$PLATFORM_EXTRA_ARGS;nqpus;$2"
 		;;
-	--nvcf-function-id)
+	--nvqc-function-id)
```
Do you expect to create an additional setting for nr gpus per virtual processor or did you intend the function-id parameter to meet this need?

1tnguyen replied:
Fixed in https://github.com/NVIDIA/cuda-quantum/commit/ae21a9b4165762420178cb7cbfa3a4cdf89892af.
Number of GPUs is now a supported option.

bettinaheim reviewed (approved):
lgtm
